src/thunderbots/software/ai/passing/pass_generator.h

```diff
@@ -0,0 +1,249 @@
+     * a lock in A, then call B, which also requires a lock, then the threads will
+     * deadlock and everything will grind to a halt.
+     *
+     */
+    class PassGenerator
+    {
+       public:
+        // Delete the default constructor, we want to force users to choose what
+        // pass quality they deem reasonable
+        PassGenerator() = delete;
+
+        // Delete the copy and assignment operators because this class really shouldn't
+        // need them and we don't want to risk doing anything nasty with the internal
+        // threading this class uses
+        PassGenerator& operator=(const PassGenerator&) = delete;
+        PassGenerator(const PassGenerator&)            = delete;
+
+        /**
+         * Create a PassGenerator with given parameters
+         *
+         * @param min_reasonable_pass_quality A value in [0,1] representing the minimum
+         *                                    quality for a pass to be considered
+         *                                    "reasonable", with higher being better
+         */
+        explicit PassGenerator(double min_reasonable_pass_quality);
+
+        /**
+         * Updates the world
+         *
+         * @param world
+         */
+        void setWorld(World world);
+
+        /**
+         * Updates the point that we are passing from
+         *
+         * @param passer_point the point we are passing from
+         */
+        void setPasserPoint(Point passer_point);
+
+        /**
+         * Set the target region that we would like to pass to
+         *
+         * @param area The area to pass to
+         */
+        void setTargetRegion(Rectangle area);
+
+        /**
+         * Gets the best pass we know of so far
+         *
+         * This only returns what we know so far. For example, if called directly after
+         * the world state is updated, it is unlikely to return good results (if any).
+         * Gradient descent must be allowed to run for some number of iterations before
+         * this can be used to get a reasonable value.
+         *
+         * @return The best currently known pass, or `std::nullopt` if there is no
+         *         reasonable pass
+         */
+        std::optional<Pass> getBestPassSoFar();
+
+        /**
+         * Destructs this PassGenerator
+         */
+        ~PassGenerator();
+
+
+       private:
+        // The number of parameters (representing a pass) that we optimize
+        // (pass_start_x, pass_start_y, pass_speed, pass_start_time)
+        static const int NUM_PARAMS_TO_OPTIMIZE = 4;
+
+
+        // Weights used to normalize the parameters that we pass to GradientDescent
+        // (see the GradientDescent documentation for details)
+        static constexpr double PASS_SPACE_WEIGHT                          = 0.01;
+        static constexpr double PASS_TIME_WEIGHT                           = 1;
+        static constexpr double PASS_SPEED_WEIGHT                          = 1;
```

Comment:
I think the ratio we used before was 1:2:6, why the change? The ratio we were using was picked using the heuristic of normalizing the range available to each parameter. So if the spatial range is [-4,4], time is [0,3], and speed is [2,6] then the weights would be 1:(8/3):2. There are probably better ways to do this though

Reply:
No particular reason, just had some random values that worked somewhat well for convergence over the static field. These will definitely get changed as part of #388 though.